### PR TITLE
Layers docs

### PR DESCRIPTION
**Pull request by gabe-l-hart**
## Description

This PR adds a new documentation page which goes in depth on the architectural layers of `caikit`!

## Notes

I was not able to get the local server running to ensure that the markdown renders into `html` cleanly. I can keep working through that, but I wanted to get the PR up since I've been squatting on this for a while.

### CODE REVIEW

File: docs/architecture_layers.md [hunk 1]
```diff
@@ -0,0 +1,977 @@
...
+
+Each sub module within `caikit.interfaces.<domain>` holds a `tasks` module that defines the key [task signatures](#tasks) for the problems to be solved in the given domain.
+
+## 5. AI Domain Libraries
+
+Within the `caikit` project, there is an evolving set of libraries that offer concrete implementations of [tasks](#tasks) based on their domains.
+
+### Caikit NLP
+
+The most mature AI domain in the `caikit` community is Natural Language Processing (NLP). The [caikit-nlp project](https://github.com/caikit/caikit-nlp) holds modules implementing the most essential tasks in NLP, particular for generative AI usecases.
```
Comment (dmuelle) [inline docs/architecture_layers.md:949]:
```suggestion
The most mature AI domain in the `caikit` community is Natural Language Processing (NLP). The [caikit-nlp project](https://github.com/caikit/caikit-nlp) holds modules implementing the most essential tasks in NLP, particularly for generative AI use cases.
```

File: docs/architecture_layers.md [hunk 2]
```diff
@@ -0,0 +1,977 @@
...
+
+* [**Text Generation**](https://github.com/caikit/caikit-nlp/tree/main/caikit_nlp/modules/text_generation): Generative models which take prompt text in and generate novel text based on the prompt.
+* [**Text Embedding**](https://github.com/caikit/caikit-nlp/tree/main/caikit_nlp/modules/text_embedding): Given input prompt text, generate embedding vectors based on a language model
+* [**Tokenization**](https://github.com/caikit/caikit-nlp/tree/main/caikit_nlp/modules/tokenization): Given a string of text, split the text into a sequence of individual tokens
+* [**Token Classification**](https://github.com/caikit/caikit-nlp/tree/main/caikit_nlp/modules/token_classification): Given a sequence of tokens, detect token groups which match certain classes (e.g. Personally Identifiable Information)
```
Comment (dmuelle) [inline docs/architecture_layers.md:954]:
```suggestion
* [**Token Classification**](https://github.com/caikit/caikit-nlp/tree/main/caikit_nlp/modules/token_classification): Given a sequence of tokens, detect token groups that match certain classes (e.g. Personally Identifiable Information)
```

File: docs/architecture_layers.md [hunk 3]
```diff
@@ -0,0 +1,977 @@
...
+
+The `caikit-computer-vision` is in its early stages. It currently holds a single `transformers` implementation of the [**Object Detection**](https://github.com/caikit/caikit-computer-vision/tree/main/caikit_computer_vision/modules/object_detection) task, but will continue to grow as usage and contributions add up.
+
+## 6. Prebuilt Runtime Images
+
+The `caikit` community, in close partnership with Red Hat's OpenShift AI team, provides several ways to build/consume container images that are ready to use in a runtime of the user's choosing.
+
+### Caikit NLP
+
+The `caikit-nlp` project provides a [Dockerfile](https://github.com/caikit/caikit-nlp/blob/main/Dockerfile) that builds an image which can be launched to start the `caikit.runtime` server(s) with the `caikit_nlp` modules configured.
```
Comment (dmuelle) [inline docs/architecture_layers.md:967]:
```suggestion
The `caikit-nlp` project provides a [Dockerfile](https://github.com/caikit/caikit-nlp/blob/main/Dockerfile) that builds an image that can be launched to start the `caikit.runtime` server(s) with the `caikit_nlp` modules configured.
```

**Review by dmuelle — COMMENTED** [review]
a few minor suggestions- feel free to ignore anything that doesnt make sense in context


At a higher level- I'm not that familiar with the territory -but- this does seem very conceptual and abstract, insular. Some of that is to be expected with an architectural doc. 

This might fall more into "nice to have", but if the hope is to orient people who want to better understand caikit- it could be helpful to some readers to have some kind of high-level example of what you might actually _do_ with this framework, or the types of advantages the architectural details afford in an example context. Some kind of through line to tie the concepts together. The document is very thorough on the _what_ and the _how_, but the _why_ could be more concrete, which might help new readers to tie everything together conceptually. Just my $0.02

File: docs/architecture_layers.md [hunk 4]
```diff
@@ -0,0 +1,977 @@
...
+title: Architecture Layers
+layout: default
+nav_order: 6
+---
+
+# Architecture Layers
+
+The `caikit` project is itself a stack of complimentary building blocks that layer on top of each other to provide rising levels of abstraction to the various functions in the space of AI problems. The layers roughly break down as follows:
+
+1. [**Useful Tools**](#1-useful-tools): At the bottom, `caikit` holds a collection of tools that are not AI-specific. These tools are useful for AI, but can be imported and used in just about any problem space where they fit.
```
[the remark this answers is not in the document]
Comment (gabe-l-hart) [reply docs/architecture_layers.md:11]: Nice, I like that clarification

File: docs/architecture_layers.md [hunk 5]
```diff
@@ -0,0 +1,977 @@
...
+# >>> }
+# >>> embeddings {
+# >>>   data: "\000\000\000\003?\346ffffff?\351\231\231\231\231\231\232?\354\314\314\314\314\314\315"
+# >>>   data_type: FLOAT
+# >>> }
+```
+
+### Data Streaming
+
+For many AI workloads, especially `training`, data volumes quickly escape the bounds of in-memory iterables. There are numerous solutions for managing iterables of data backed by disk or even remote services. In `caikit`, these are all managed via the [DataStream](https://github.com/caikit/caikit/blob/main/caikit/core/data_model/streams/data_stream.py) utility. This class provides consistent iteration semantics wrapping arbitrary [python generators](https://wiki.python.org/moin/Generators).
```
[the remark this answers is not in the document]
Comment (gabe-l-hart) [reply docs/architecture_layers.md:385]: Good catch. The former!

**Comment by gabe-l-hart** [comment]
Thanks for the eagle-eyed review @dmuelle! I pushed updates for all of them (mostly your suggestions with a few further tweaks).

**Review by dmuelle — APPROVED** [review]
LGTM, thanks

**Pull request merged**